File: online_radyo_flutter/lib/screens/radio_player_screen.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:just_audio/just_audio.dart';

import '../models/radio_room.dart';
import '../network/radio_socket_manager.dart';

class RadioPlayerScreen extends StatefulWidget {
  final RadioRoom room;

  const RadioPlayerScreen({
    super.key,
    required this.room,
  });

  @override
  State<RadioPlayerScreen> createState() => _RadioPlayerScreenState();
}

class _RadioPlayerScreenState extends State<RadioPlayerScreen> {
  final RadioSocketManager socket = RadioSocketManager.instance;
  final AudioPlayer player = AudioPlayer();

  Timer? syncTimer;

  String musicTitle = "Çalan müzik bekleniyor...";
  String statusText = "Odaya bağlanılıyor...";
  String? currentMusicUrl;

  @override
  void initState() {
    super.initState();

    setupSocket();
    socket.joinRoom(widget.room.id);

    syncTimer = Timer.periodic(const Duration(seconds: 5), (_) {
      socket.requestSync(widget.room.id);
    });
  }

  void setupSocket() {
    socket.onMessage = (message) {
      handleSocketMessage(message);
    };

    socket.onError = (error) {
      setState(() {
        statusText = "Bağlantı hatası: $error";
      });
    };

    socket.connect();
  }

  Future<void> handleSocketMessage(String message) async {
    final Map<String, dynamic> json = jsonDecode(message);
    final String type = json["type"] ?? "";

    if (type == "PLAYBACK_STATE") {
      final int incomingRoomId = json["roomId"] ?? -1;

      if (incomingRoomId != widget.room.id) return;

      final String title = json["title"] ?? "Bilinmeyen müzik";
      final String musicUrl = json["musicUrl"] ?? "";
      final double positionSeconds =
      (json["positionSeconds"] ?? 0).toDouble();

      await playOrSyncMusic(
        title: title,
        musicUrl: musicUrl,
        positionSeconds: positionSeconds,
      );
    }

    else if (type == "NO_MUSIC") {
      await player.pause();

      setState(() {
        musicTitle = "Bu odada şu an müzik yok";
        statusText = "Bekleniyor...";
      });
    }
  }

  Future<void> playOrSyncMusic({
    required String title,
    required String musicUrl,
    required double positionSeconds,
  }) async {
    setState(() {
      musicTitle = title;
      statusText = "Dinleniyor...";
    });

    final Duration targetPosition =
    Duration(milliseconds: (positionSeconds * 1000).toInt());

    if (currentMusicUrl != musicUrl) {
      currentMusicUrl = musicUrl;

      await player.setUrl(musicUrl);
      await player.seek(targetPosition);
      await player.play();
      return;
    }

    final Duration? currentPosition = player.position;
    final int diffMs =
    (currentPosition!.inMilliseconds - targetPosition.inMilliseconds).abs();

    if (diffMs > 1200) {
      await player.seek(targetPosition);
    }

    if (!player.playing) {
      await player.play();
    }
  }

  @override
  void dispose() {
    syncTimer?.cancel();
    player.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.room.roomName),
      ),
      body: Padding(
        padding: const EdgeInsets.all(20),
        child: Column(
          children: [
            Text(
              widget.room.roomName,
              style: Theme.of(context).textTheme.headlineMedium,
              textAlign: TextAlign.center,
            ),

            const SizedBox(height: 16),

            Text(
              musicTitle,
              style: Theme.of(context).textTheme.titleLarge,
              textAlign: TextAlign.center,
            ),

            const SizedBox(height: 8),

            Text(
              statusText,
              textAlign: TextAlign.center,
            ),

            const Spacer(),

            const Icon(
              Icons.radio,
              size: 90,
            ),

            const SizedBox(height: 16),

            const Text(
              "Bu ekran sadece dinleyici modudur.",
              textAlign: TextAlign.center,
            ),

            const Spacer(),
          ],
        ),
      ),
    );
  }
}

File: OnlineRadioJetpack/app/src/main/java/com/example/onlineradiojetpack/ui/RadioPlayerScreen.kt
package com.example.onlineradiojetpack.ui

import android.os.Handler
import android.os.Looper
import android.view.ViewGroup
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.viewinterop.AndroidView
import androidx.compose.ui.unit.dp
import androidx.media3.common.MediaItem
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.ui.PlayerView
import com.example.onlineradiojetpack.network.RadioSocketManager
import kotlinx.coroutines.delay
import org.json.JSONObject
import kotlin.math.abs

@Composable
fun RadioPlayerScreen(
    roomId: Int,
    roomName: String
) {
    val context = LocalContext.current

    val mainHandler = remember {
        Handler(Looper.getMainLooper())
    }

    var musicTitle by remember { mutableStateOf("Çalan müzik bekleniyor...") }
    var status by remember { mutableStateOf("Odaya bağlanılıyor...") }
    var currentMusicUrl by remember { mutableStateOf<String?>(null) }

    val player = remember {
        ExoPlayer.Builder(context).build()
    }

    fun playOrSyncMusic(
        title: String,
        musicUrl: String,
        positionSeconds: Double
    ) {
        musicTitle = title
        status = "Dinleniyor..."

        val targetPositionMs = (positionSeconds * 1000).toLong()

        if (currentMusicUrl != musicUrl) {
            currentMusicUrl = musicUrl

            val mediaItem = MediaItem.fromUri(musicUrl)

            player.setMediaItem(mediaItem)
            player.prepare()
            player.seekTo(targetPositionMs)
            player.play()
            return
        }

        val diff = abs(player.currentPosition - targetPositionMs)

        if (diff > 1200) {
            player.seekTo(targetPositionMs)
        }

        if (!player.isPlaying) {
            player.play()
        }
    }

    LaunchedEffect(roomId) {
        RadioSocketManager.onMessage = { message ->

            mainHandler.post {
                try {
                    val json = JSONObject(message)
                    val type = json.optString("type")

                    when (type) {
                        "PLAYBACK_STATE" -> {
                            val incomingRoomId = json.optInt("roomId")

                            if (incomingRoomId == roomId) {
                                playOrSyncMusic(
                                    title = json.optString("title"),
                                    musicUrl = json.optString("musicUrl"),
                                    positionSeconds = json.optDouble("positionSeconds", 0.0)
                                )
                            }
                        }

                        "NO_MUSIC" -> {
                            musicTitle = "Bu odada şu an müzik yok"
                            status = "Bekleniyor..."
                            player.pause()
                        }
                    }

                } catch (e: Exception) {
                    status = "Mesaj işleme hatası: ${e.message}"
                }
            }
        }

        RadioSocketManager.onError = { error ->
            mainHandler.post {
                status = "Bağlantı hatası: $error"
            }
        }

        RadioSocketManager.connect()
        RadioSocketManager.joinRoom(roomId)

        while (true) {
            delay(5000)
            RadioSocketManager.requestSync(roomId)
        }
    }

    DisposableEffect(Unit) {
        onDispose {
            mainHandler.post {
                player.release()
            }
        }
    }

    Scaffold { padding ->
        Column(
            modifier = Modifier
                .padding(padding)
                .fillMaxSize()
                .padding(20.dp)
        ) {
            Text(
                text = roomName,
                style = MaterialTheme.typography.headlineMedium
            )

            Spacer(modifier = Modifier.height(12.dp))

            Text(
                text = musicTitle,
                style = MaterialTheme.typography.titleMedium
            )

            Spacer(modifier = Modifier.height(8.dp))

            Text(
                text = status,
                style = MaterialTheme.typography.bodyMedium
            )

            Spacer(modifier = Modifier.height(16.dp))

            AndroidView(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(90.dp),
                factory = {
                    PlayerView(it).apply {
                        this.player = player
                        useController = false
                        layoutParams = ViewGroup.LayoutParams(
                            ViewGroup.LayoutParams.MATCH_PARENT,
                            ViewGroup.LayoutParams.WRAP_CONTENT
                        )
                    }
                }
            )
        }
    }
}